### PR TITLE
feat(ui): add `ui-components` package

### PR DESCRIPTION
**Pull request by avivkeller**
This PR moves the first few components into the `ui-components` package.

For now, only the `Icons/` folder has been moved, as it's quite simple to move (unlike some other components), and it gives a good basis of what this package will look like.

Over coming PRs, I'll slowly move more pieces into this package.

### CODE REVIEW

**Comment by RedYetiDev** [comment]
Currently, this moves **a lot** of components to this new pacakge. Personally, I think that a better way to do this is incrementally. As in, this PR should move 1-2 components, and the rest should slowly be moved over. That way, if anything breaks / changes, it'll be very easy to see what exactly happened. 


If you agree, feel free to :+1: this message, and I'll make the changes.

**Comment by avivkeller** [comment]
Any idea why Vercel is failing?

**Comment by AugustinMauroy** [comment]
Maybe create a package stories ?

**Comment by AugustinMauroy** [comment]
> Any idea why Vercel is failing?

cc @ovflowd and @bmuenzenmeyer

File: apps/site/components/Containers/NavBar/index.tsx [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 import Hamburger from '@heroicons/react/24/solid/Bars3Icon';
 import XMark from '@heroicons/react/24/solid/XMarkIcon';
+import GitHub from '@node-core/ui-components/Icons/Social/GitHub';
```
Comment (ovflowd) [inline apps/site/components/Containers/NavBar/index.tsx]: I think we should suffix all icon imports with the word "Icon" (including these other from heroIcons)

Comment (AugustinMauroy) [reply apps/site/components/Containers/NavBar/index.tsx]: +1 for that !

Comment (avivkeller) [reply apps/site/components/Containers/NavBar/index.tsx]: Shouldn't this be done in a follow-up PR, as to minimize the diff?

**Comment by avivkeller** [comment]
FYI I'm drafting until the Next.js update is done.

Feel free to review, this just can't land until that does (given that it modifies the `package-lock.json` files along with other configuration/storybook files, which is also being done by that more important upgrade).

File: packages/ui-components/tailwind.config.ts
Comment (AugustinMauroy) [inline packages/ui-components/tailwind.config.ts:1]: may be create a shared tailwind config ? to avoid duplicate code

**Comment by avivkeller** [comment]
Hey, can someone let me know why Vercel is failing?

File: apps/site/.storybook/preview.tsx [hunk 2]
```diff
@@ -3,9 +3,13 @@ import { withThemeByDataAttribute } from '@storybook/addon-themes';
 import type { Preview, ReactRenderer } from '@storybook/react';
 import { NextIntlClientProvider } from 'next-intl';
 
-import { STORYBOOK_MODES, STORYBOOK_SIZES } from '@/.storybook/constants';
 import { NotificationProvider } from '@/providers/notificationProvider';
 
+import {
+  STORYBOOK_MODES,
+  STORYBOOK_SIZES,
+} from '../../../.storybook/constants';
```
Comment (ovflowd) [inline apps/site/.storybook/preview.tsx:8]: FYI Storybook should exist only within the UI components package :) -- not within the root of the repository.

Comment (avivkeller) [reply apps/site/.storybook/preview.tsx:8]: Storybook is in use in both the UI components, and the `app/site`. To prevent duplicate code, I moved it to the repo root (like Tailwind), what do you have in mind?

Comment (ovflowd) [reply apps/site/.storybook/preview.tsx:8]: Which components are on app/site that use Storybook? IMO, Storybook should **ONLY** exist for the UI components package.

Because semantically, that's what it is supposed to be used for. For describing components. I think we use it on apps/site only for visual regression, for whose I don't think they should be storybooks but just snapshots. We should probably make that change before we make this package then.

Comment (ovflowd) [reply apps/site/.storybook/preview.tsx:8]: We can still use Chromatic for that: https://www.chromatic.com/features/visual-test.

IMO you can do that on this PR, decoupling stories from website-only components (which actually are combination of the base components) and simply use Cypress or Node.js or whatever for the visual regression tests with Chromatic.

File: apps/site/app/[locale]/next-data/og/[category]/[title]/route.tsx [hunk 3]
```diff
@@ -1,7 +1,7 @@
+import HexagonGrid from '@node-core/ui-components/Icons/HexagonGrid';
```
Comment (ovflowd) [inline apps/site/app/[locale]/next-data/og/[category]/[title]/route.tsx:1]: Lets suffix icons with the word Icon

File: apps/site/components/Common/Select/index.stories.tsx [hunk 4]
```diff
@@ -1,9 +1,9 @@
+import Apple from '@node-core/ui-components/Icons/Platform/Apple';
```
Comment (ovflowd) [inline apps/site/components/Common/Select/index.stories.tsx]: Icons should be suffixed

File: apps/site/components/Containers/NavBar/index.tsx (hunk 1 above)
Comment (ovflowd) [reply apps/site/components/Containers/NavBar/index.tsx]: The diff is already large enough, and IMO, it is just the import that you would be updating, which already affects this PR, so IMO there wouldn't be extra diff besides of usage updates.

File: apps/site/tailwind.config.ts [hunk 5]
```diff
@@ -1,161 +1 @@
...
-      },
-    },
-  },
-  darkMode: ['selector', '[data-theme="dark"]'],
-  plugins: [
-    require('@savvywombat/tailwindcss-grid-areas'),
-    require('@tailwindcss/container-queries'),
-  ],
-} satisfies Config;
+export { default } from '../../tailwind.config';
```
Comment (ovflowd) [inline apps/site/tailwind.config.ts]: IMO tailwind config should be part of the UI components package as a base tailwind config and we should be able to import it on the Website package.

Comment (ovflowd) [reply apps/site/tailwind.config.ts]: This also means that font loading should be done within the UI components package instead on Next.js-side.

Meaning, we should stop using `next-fonts` and use something that is pure react-ish and that can integrate natively with Tailwind.

Comment (ovflowd) [reply apps/site/tailwind.config.ts]: Meaning this package (ui-components) should also export a base CSS file that is something like:

```css
/* global.css */
@tailwind base;
@tailwind components;
@tailwind utilities;

/* update this to use the two fonts we use */
@import url("https://fonts.googleapis.com/css2?family=Poppins:wght@400;500&display=swap");
```

File: packages/ui-components/.storybook/preview-head.html [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<link rel="preconnect" href="https://fonts.googleapis.com" />
+<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
+
+<link
```
Comment (ovflowd) [inline packages/ui-components/.storybook/preview-head.html]: https://github.com/nodejs/nodejs.org/pull/7223/files#r1846684448

File: packages/ui-components/.storybook/main.ts
Comment (ovflowd) [inline packages/ui-components/.storybook/main.ts:1]: https://github.com/nodejs/nodejs.org/pull/7223/files#r1845414033

File: tailwind.config.ts
Comment (ovflowd) [inline tailwind.config.ts:1]: https://github.com/nodejs/nodejs.org/pull/7223/files#r1846680914

**Review by ovflowd — COMMENTED** [review]
I left an initial round of comments 🙏

File: apps/site/tailwind.config.ts (hunk 5 above)
Comment (AugustinMauroy) [reply apps/site/tailwind.config.ts]: Wow, I'm not a fan of what you're proposing. Why does `next-font` exist, it's to prevent the user of the page from depending on google. but if we do what you suggest that's what's going to happen. Our user will have to go through google. We would therefore need to add a file download system to the building time. To note that `googleapis` is blocked by not badly of proxy http in company or other thus this user would not have the font.

File: apps/site/.storybook/preview.tsx (hunk 2 above)
Comment (avivkeller) [reply apps/site/.storybook/preview.tsx:8]: I agree with this overall, I just don't think that this should be done _now_. Currently, while only a subset of features are available in the UI package, I think it's okay to share a storybook config, as both the UI package and the apps/site package ned it. Later on, once more stuff is moved into the package, the storybook should only exist there, just not yet. Not while only the icons have been transferred.

But, in the future, this is probably the best direction to go in.

File: apps/site/tailwind.config.ts (hunk 5 above)
Comment (avivkeller) [reply apps/site/tailwind.config.ts]: I agree with @AugustinMauroy. It's not ideal for the users to need to fetch the font from an external source. If we _really_ wanted to get rid of `next-font`, we could store the font file as an asset, but even if that was the direction we were going to do in, it should be it's own discussion, and not part of this PR (IMO).

Comment (AugustinMauroy) [reply apps/site/tailwind.config.ts]: >  it should be it's own discussion, and not part of this PR (IMO).

So I think we need to find a solution for this that doesn't depend on Google and then look for the most refined solution.

**Comment by avivkeller** [comment]
Hey @ovflowd. I've responded to some of your concerns, but I still have my reservations about the storybook / tailwind concerns, PTAL when you get a chance, thanks!

File: apps/site/.storybook/preview.tsx (hunk 2 above)
Comment (ovflowd) [reply apps/site/.storybook/preview.tsx:8]: I get it, I just believe this gives an excellent opportunity for so; This PR is already going to be huge, and I would say that ensuring that Storybooks only exist for actual UI components (instead of composite components/fragments/sections of a page) and that we still ensure visual regression on everything is definitely part of the goal of this package IMO.

We don't need to rush here, so if the issue is time, take your time :)

File: apps/site/tailwind.config.ts (hunk 5 above)
Comment (ovflowd) [reply apps/site/tailwind.config.ts]: It is important to mention that all that next-fonts does is download the font during build-time and bundle it within our assets. It still imports the font the same way with a `@import url`

Comment (ovflowd) [reply apps/site/tailwind.config.ts]: There are a couple of issues with next-fonts and why it doesn't work for us:

- We cannot use `net-fonts` in the UI components package. That package is agnostic and should only require TailwindCSS and React.
  - The fonts are essential for the UI components; without the right font families, our Storybooks and Visual Regression tests will have no meaning.
- We shouldn't rely on Next.js to have custom fonts within our Application. `next-fonts` is just an easy one-liner abstraction that is nice for most simple projects, but not here.

>  Our user will have to go through google. 

Our servers will anyway need to go through Google, and Google Fonts is also used everywhere across the internet. The chances that these fonts will already be cached by the time they need to request them from our website (at least Open Sans) reduce already the extra request they need to make to our infrastructure to load the font.

> Why does next-font exist, it's to prevent the user of the page from

Just because something exists doesn't mean we need to use it.

**Comment by ovflowd** [comment]
Thanks @RedYetiDev I've replied to your points ;)

File: apps/site/.storybook/preview.tsx (hunk 2 above)
Comment (avivkeller) [reply apps/site/.storybook/preview.tsx:8]: The current Storybook configuration in `/apps/site` differs from the one in `packages/ui-components` because the `ui-components` package doesn't rely on Next.js internally. If we were to move the Storybook setup into `ui-components` now, we'd have to introduce Next.js dependencies, as the configuration in `apps/site` depends on Next.js internals. The goal of this package is to _not_ have Next.js dependencies.

A more effective approach IMO would be to gradually migrate components to the `packages/ui-components` package while removing Next.js-related elements from their Storybook configurations. Once all configurations are free of Next.js and consolidated in `packages/ui-components`, transferring the entire Storybook setup to this package will be much smoother.

By "Next.js-free," I mean the removal of dependencies like `next-intl`, `Link`, and similar Next.js-specific elements.

File: apps/site/tailwind.config.ts (hunk 5 above)
Comment (avivkeller) [reply apps/site/tailwind.config.ts]: I misunderstood what you were initially asking. I thought you meant to get rid of next-fonts for the entire site. Now I understand that you mean only for this package. Yes, you are right, I'll update the CSS file.

**Comment by avivkeller** [comment]
The deployment was successful!!

**Comment by ovflowd** [comment]
@RedYetiDev maybe you're unfamiliar with the website workflows, but we add a `github_actions:pull-request` label to trigger our CI xD

File: apps/site/.storybook/preview.tsx (hunk 2 above)
Comment (ovflowd) [reply apps/site/.storybook/preview.tsx:8]: > The current Storybook configuration in /apps/site differs from the one in packages/ui-components because the ui-components package doesn't rely on Next.js internally.

That doesn't make sense. The current Storybook configuration is Next.js-agnostic.

Comment (ovflowd) [reply apps/site/.storybook/preview.tsx:8]: > If we were to move the Storybook setup into ui-components now, we'd have to introduce Next.js dependencies, as the configuration in apps/site depends on Next.js internals. The goal of this package is to not have Next.js dependencies.

Could you elaborate where you saw that? I'm pretty sure it is not.

Comment (ovflowd) [reply apps/site/.storybook/preview.tsx:8]: > By "Next.js-free," I mean the removal of dependencies like next-intl, Link, and similar Next.js-specific elements.

You do know these lines are to "mock" components that use such dependencies, and that the ui-components library does not use these said dependencies, since the Link component for example, will now use Dependency Injection... Also, the Storybook should only exist for UI components, meaning there is nothing Next.js specific; I'm not sure where the confusion is here.

Sorry, but I'm adamant that Storybook should only be for ui-components, I won't approve this PR without that :/

Comment (ovflowd) [reply apps/site/.storybook/preview.tsx:8]: > A more effective approach IMO would be to gradually migrate components to the packages/ui-components package while removing Next.js-related elements from their Storybook configurations. Once all configurations are free of Next.js and consolidated in packages/ui-components, transferring the entire Storybook setup to this package will be much smoother.

I don't think we need a gradative migration on the main branch; as mentioned before, nothing stops you from making a feature branch and gradatively migrating the components; there's no rush to getting this done, and I doubt that conflicts will arise over time...

I'll defer to the rest of the @nodejs/nodejs-website if they are fine with merging this, as it will complicate things further with two different Storybooks.

I only want that we have a dedicated Storybook for UI components and that's it; No other Storybook, not mess needed. I won't block this PR, but I'm expressing my discontent.

File: packages/ui-components/.storybook/tailwind.css [hunk 7]
```diff
@@ -0,0 +1,7 @@
+@import 'https://fonts.googleapis.com/css2?family=IBM+Plex+Mono:ital,wght@0,100;0,200;0,300;0,400;0,500;0,600;0,700;1,100;1,200;1,300;1,400;1,500;1,600;1,700&family=Open+Sans:ital,wght@0,300..800;1,300..800&display=swap';
```
Comment (ovflowd) [inline packages/ui-components/.storybook/tailwind.css:1]: FYI don't forget to add a preconnect on the Next.js base layout.

File: packages/ui-components/.storybook/tailwind.css [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+@import 'tailwindcss/base';
+@import 'tailwindcss/components';
+@import 'tailwindcss/utilities';
+@import '../../../apps/site/styles/base.css';
```
Comment (ovflowd) [inline packages/ui-components/.storybook/tailwind.css:5]: It feels weired that the ui components package depends on another package's styles (the website) -- that should definitely not be the case. These styles should live here if nothing else.

Comment (ovflowd) [reply packages/ui-components/.storybook/tailwind.css:5]: (But I see this is only for storybook, so 🤷)

File: apps/site/tailwind.config.ts (hunk 5 above)
Comment (ovflowd) [reply apps/site/tailwind.config.ts]: I actually did mean for everything. There's no point of having next-fonts if the ui-components package already loads the fonts 🤷

Comment (ovflowd) [reply apps/site/tailwind.config.ts]: Otherwise we are double loading fonts.

**Comment by avivkeller** [comment]
Crap! Rebase issues!

**Comment by avivkeller** [comment]
Per <https://openjs-foundation.slack.com/archives/CVAMEJ4UV/p1736551106164799?thread_ts=1736546605.687069&cid=CVAMEJ4UV>.

TL;DR a botched rebase messed up the entire tree (and sent it back ~200 commits). Tomorrow I'll make a new feature branch with the correct tree.

**Comment by ovflowd** [comment]
> Per https://openjs-foundation.slack.com/archives/CVAMEJ4UV/p1736551106164799?thread_ts=1736546605.687069&cid=CVAMEJ4UV.
> 
> TL;DR a botched rebase messed up the entire tree (and sent it back ~200 commits). Tomorrow I'll make a new feature branch with the correct tree.

Oh no, I bet this will take lots of work :(

**Comment by avivkeller** [comment]
I actually managed to get it done pretty quickly (yay!)

https://github.com/nodejs/nodejs.org/pull/7401.
